ludwig/features/audio_feature.py

```diff
@@ -0,0 +1,301 @@
+        audio_feature_dict = preprocessing_parameters['audio_feature']
+        first_audio_file_path = column[0]
+        _, sampling_rate_in_hz = soundfile.read(first_audio_file_path)
+
+        feature_dim = AudioBaseFeature._get_feature_dim(audio_feature_dict, sampling_rate_in_hz)
+        audio_file_length_limit_in_s = preprocessing_parameters['audio_file_length_limit_in_s']
+        max_length = AudioBaseFeature._get_max_length_feature(audio_feature_dict, sampling_rate_in_hz, audio_file_length_limit_in_s)
+        return { 
+                'feature_dim': feature_dim,
+                'sampling_rate_in_hz': sampling_rate_in_hz,
+                'max_length': max_length
+               }
+
+    @staticmethod
+    def _get_feature_dim(audio_feature_dict, sampling_rate_in_hz):
+        feature_type = audio_feature_dict['type']
+
+        if(feature_type == 'raw'):
+            feature_dim = 1
+        elif(feature_type == 'stft_phase'):
+            feature_dim_symmetric = get_length_in_samp(audio_feature_dict['window_length_in_s'], sampling_rate_in_hz)
+            feature_dim = 2 * get_non_symmetric_length(feature_dim_symmetric)
+        elif(feature_type in ['stft', 'group_delay']):
+            feature_dim_symmetric = get_length_in_samp(audio_feature_dict['window_length_in_s'], sampling_rate_in_hz)
+            feature_dim = get_non_symmetric_length(feature_dim_symmetric)
+        else:
+            raise ValueError('{} is not recognized.'.format(feature_type))
+
+        return feature_dim
+
+    @staticmethod
+    def _read_audio_and_transform_to_feature(filepath, audio_feature_dict, feature_dim, max_length, padding_value, normalization_type, audio_stats):
+        """
+        :param filepath: path to the audio
+        :param audio_feature_dict: dictionary describing audio feature see default
+        :param feature_dim: dimension of each feature frame
+        :param max_length: max audio length defined by user in samples
+        """
+
+        feature_type = audio_feature_dict['type']
+        audio, sampling_rate_in_hz = soundfile.read(filepath)
+        AudioBaseFeature._update(audio_stats, audio, sampling_rate_in_hz)
+
+        if(feature_type == 'raw'):
+            audio_feature = np.expand_dims(audio, axis=-1)
+        elif(feature_type in ['stft', 'stft_phase', 'group_delay']):
+            audio_feature = np.transpose(AudioBaseFeature._get_2D_feature(audio, feature_type, audio_feature_dict, sampling_rate_in_hz))
+        else:
+            raise ValueError('{} is not recognized.'.format(feature_type))
+
+        feature_length = audio_feature.shape[0]
+        broadcast_feature_length = min(feature_length, max_length)
+        audio_feature_padded = np.full((max_length, feature_dim), padding_value, dtype=np.float32)
+        audio_feature_padded[:broadcast_feature_length, :] = audio_feature[:max_length, :]
+        return audio_feature_padded
+
+    @staticmethod
+    def _update(audio_stats, audio, sampling_rate_in_hz):
```

Comment:
Maybe call it _update_stats or something more informative?

Reply:
1cf2433